File: cleanup_vpc.js
const { EC2Client, DeleteVpcCommand, DescribeVpcEndpointsCommand, DeleteVpcEndpointsCommand, DescribeRouteTablesCommand, DeleteRouteTableCommand, DeleteInternetGatewayCommand, DetachInternetGatewayCommand, DeleteSubnetCommand, DeleteSecurityGroupCommand, DescribeVpcsCommand, DescribeInternetGatewaysCommand, DescribeSubnetsCommand, DescribeSecurityGroupsCommand, DescribeNetworkInterfacesCommand, DeleteNetworkInterfaceCommand } = require("@aws-sdk/client-ec2");

const client = new EC2Client({ region: process.argv[3] });
const vpcId = process.argv[2];

async function cleanupVpc() {
    console.log(`Starting cleanup for VPC ${vpcId}...`);

    try {
        // 1. Delete VPC Endpoints
        const endpoints = await client.send(new DescribeVpcEndpointsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }));
        if (endpoints.VpcEndpoints.length > 0) {
            const ids = endpoints.VpcEndpoints.map(e => e.VpcEndpointId);
            await client.send(new DeleteVpcEndpointsCommand({ VpcEndpointIds: ids }));
            console.log(`Deleted endpoints: ${ids.join(', ')}`);
        }

        // 2. Delete Internet Gateways
        const igws = await client.send(new DescribeInternetGatewaysCommand({ Filters: [{ Name: 'attachment.vpc-id', Values: [vpcId] }] }));
        for (const igw of igws.InternetGateways) {
            await client.send(new DetachInternetGatewayCommand({ InternetGatewayId: igw.InternetGatewayId, VpcId: vpcId }));
            await client.send(new DeleteInternetGatewayCommand({ InternetGatewayId: igw.InternetGatewayId }));
            console.log(`Deleted IGW: ${igw.InternetGatewayId}`);
        }

        // 3. Delete Subnets (and interfaces)
        const subnets = await client.send(new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }));
        for (const subnet of subnets.Subnets) {
            const interfaces = await client.send(new DescribeNetworkInterfacesCommand({ Filters: [{ Name: 'subnet-id', Values: [subnet.SubnetId] }] }));
            for (const iface of interfaces.NetworkInterfaces) {
                await client.send(new DeleteNetworkInterfaceCommand({ NetworkInterfaceId: iface.NetworkInterfaceId }));
                console.log(`Deleted Interface: ${iface.NetworkInterfaceId}`);
            }
            await client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
            console.log(`Deleted Subnet: ${subnet.SubnetId}`);
        }

        // 4. Delete Route Tables (non-main)
        const rts = await client.send(new DescribeRouteTablesCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }));
        for (const rt of rts.RouteTables) {
            const isMain = rt.Associations.some(a => a.Main);
            if (!isMain) {
                // Dissociate first if needed (usually auto-dissociated when subnet deleted)
                await client.send(new DeleteRouteTableCommand({ RouteTableId: rt.RouteTableId }));
                console.log(`Deleted Route Table: ${rt.RouteTableId}`);
            }
        }

        // 5. Delete Security Groups (non-default)
        const sgs = await client.send(new DescribeSecurityGroupsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }));
        for (const sg of sgs.SecurityGroups) {
            if (sg.GroupName !== 'default') {
                try {
                    await client.send(new DeleteSecurityGroupCommand({ GroupId: sg.GroupId }));
                    console.log(`Deleted Security Group: ${sg.GroupId}`);
                } catch (e) {
                    console.log(`Retrying SG delete later: ${sg.GroupId}`);
                }
            }
        }

        // 6. Delete VPC
        await client.send(new DeleteVpcCommand({ VpcId: vpcId }));
        console.log(`Successfully deleted VPC: ${vpcId}`);

    } catch (error) {
        console.error("Error cleaning up VPC:", error);
        process.exit(1);
    }
}

cleanupVpc();

File: cleanup_vpc.py
import boto3
import sys

def delete_vpc(vpc_id, region):
    ec2 = boto3.resource('ec2', region_name=region)
    ec2_client = boto3.client('ec2', region_name=region)
    vpc = ec2.Vpc(vpc_id)

    print(f"Starting cleanup for VPC {vpc_id}...")

    # Delete endpoints
    for ep in vpc_endpoints = ec2_client.describe_vpc_endpoints(Filters=[{'Name': 'vpc-id', 'Values': [vpc_id]}])['VpcEndpoints']:
        ec2_client.delete_vpc_endpoints(VpcEndpointIds=[ep['VpcEndpointId']])
        print(f"Deleted VPC Endpoint: {ep['VpcEndpointId']}")

    # Delete peering connections
    for pcx in vpc.accepted_vpc_peering_connections.all():
        pcx.delete()
        print(f"Deleted Peering Connection: {pcx.id}")
    for pcx in vpc.requested_vpc_peering_connections.all():
        pcx.delete()
        print(f"Deleted Peering Connection: {pcx.id}")

    # Delete route tables (except main)
    for rt in vpc.route_tables.all():
        is_main = False
        for assoc in rt.associations:
            if assoc.main:
                is_main = True
                break
        if not is_main:
            for assoc in rt.associations:
                assoc.delete()
            rt.delete()
            print(f"Deleted Route Table: {rt.id}")

    # Delete internet gateways
    for igw in vpc.internet_gateways.all():
        vpc.detach_internet_gateway(InternetGatewayId=igw.id)
        igw.delete()
        print(f"Deleted Internet Gateway: {igw.id}")

    # Delete subnets
    for subnet in vpc.subnets.all():
        # Delete interfaces in subnet
        for interface in subnet.network_interfaces.all():
            interface.delete()
            print(f"Deleted Network Interface: {interface.id}")
        subnet.delete()
        print(f"Deleted Subnet: {subnet.id}")

    # Delete security groups (except default)
    for sg in vpc.security_groups.all():
        if sg.group_name != 'default':
            sg.delete()
            print(f"Deleted Security Group: {sg.id}")

    # Delete VPC
    ec2_client.delete_vpc(VpcId=vpc_id)
    print(f"Successfully deleted VPC: {vpc_id}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python cleanup_vpc.py <vpc_id> <region>")
        sys.exit(1)
    
    delete_vpc(sys.argv[1], sys.argv[2])
